src/auth/models.cpp

```diff
@@ -282,8 +305,34 @@ FineGrainedAccessPermissions FineGrainedAccessPermissions::Deserialize(const nlo
   return FineGrainedAccessPermissions(data["grants"], data["denies"]);
 }
 
-const std::unordered_set<std::string> &FineGrainedAccessPermissions::grants() const { return grants_; }
-const std::unordered_set<std::string> &FineGrainedAccessPermissions::denies() const { return denies_; }
+const std::unordered_map<std::string, uint64_t> &FineGrainedAccessPermissions::grants() const { return grants_; }
+const std::unordered_map<std::string, uint64_t> &FineGrainedAccessPermissions::denies() const { return denies_; }
+
+void FineGrainedAccessPermissions::grant_(const std::string &permission, LabelPermission label_permission) {
+  uint64_t shift = 1;
+  uint64_t perm = 0;
+  auto u_label_perm = static_cast<uint64_t>(label_permission);
+
+  while (u_label_perm != 0) {
+    perm |= u_label_perm;
+    u_label_perm >>= shift;
+  }
+
+  grants_[permission] = perm;
+}
+
+void FineGrainedAccessPermissions::deny_(const std::string &permission, LabelPermission label_permission) {
```

Comment:
is this common to have underscore at the end

Reply:
No naming fixes are already made